lightwallet-client-lib/src/main/java/co/electriccoin/lightwallet/client/internal/CoroutineLightWalletClientImpl.kt

```diff
@@ -0,0 +1,200 @@
+        }
+    }
+
+    override suspend fun fetchTransaction(txId: ByteArray): Response<RawTransactionUnsafe> {
+        if (txId.isEmpty()) {
+            return Response.Failure.Client.NullIdTransaction()
+        }
+
+        return try {
+            val request = Service.TxFilter.newBuilder().setHash(ByteString.copyFrom(txId)).build()
+
+            val response = requireChannel().createStub().getTransaction(request)
+
+            val transactionResponse = RawTransactionUnsafe.new(response)
+
+            Response.Success(transactionResponse)
+        } catch (e: StatusRuntimeException) {
+            GrpcStatusResolver.resolveFailureFromStatus(e)
+        }
+    }
+
+    override suspend fun fetchUtxos(
+        tAddress: String,
+        startHeight: BlockHeightUnsafe
+    ): Flow<Service.GetAddressUtxosReply> {
+        if (tAddress.isBlank()) {
+            return emptyFlow()
+        }
+        return requireChannel().createStub().getAddressUtxosStream(
+            Service.GetAddressUtxosArg.newBuilder().setAddress(tAddress)
+                .setStartHeight(startHeight.value).build()
+        )
+    }
+
+    override fun getTAddressTransactions(
+        tAddress: String,
+        blockHeightRange: ClosedRange<BlockHeightUnsafe>
+    ): Flow<Service.RawTransaction> {
+        if (blockHeightRange.isEmpty() || tAddress.isBlank()) {
+            return emptyFlow()
+        }
+        return requireChannel().createStub().getTaddressTxids(
+            Service.TransparentAddressBlockFilter.newBuilder().setAddress(tAddress)
+                .setRange(blockHeightRange.toBlockRange()).build()
+        )
+    }
+
+    override fun shutdown() {
+        channel.shutdown()
+    }
+
+    override fun reconnect() {
+        channel.shutdown()
+        channel = channelFactory.newChannel(lightWalletEndpoint)
+    }
+
+    // test code
+    internal var stateCount = 0
+    internal var state: ConnectivityState? = null
+    private fun requireChannel(): ManagedChannel {
+        state = channel.getState(false).let { new ->
+            if (state == new) stateCount++ else stateCount = 0
+            new
+        }
+        channel.resetConnectBackoff()
+        // twig(
+        //     "getting channel isShutdown: ${channel.isShutdown}  " +
+        //         "isTerminated: ${channel.isTerminated} " +
+        //         "getState: $state stateCount: $stateCount",
+        //     -1
+        // )
+        return channel
+    }
+
+    companion object {
+        fun new(
+            channelFactory: ChannelFactory,
+            lightWalletEndpoint: LightWalletEndpoint
+        ): CoroutineLightWalletClientImpl {
+            return CoroutineLightWalletClientImpl(channelFactory, lightWalletEndpoint)
+        }
+    }
+}
+
+private fun Channel.createStub(timeoutSec: Duration = 60.seconds) =
+    CompactTxStreamerGrpcKt.CompactTxStreamerCoroutineStub(this, CallOptions.DEFAULT)
+        .withDeadlineAfter(timeoutSec.inWholeSeconds, TimeUnit.SECONDS)
```

Comment:
Please verify that this is the behavior you desire. I think that `withDeadlineAfter()` will shut down the channel after the given seconds regardless of it actually being transmitting in a very responsive communication.

Reply:
Thanks Pacu, yes it will close it. So if I understand it right, you're saying that we should have this function as is in the blocking version of the client, but not in the coroutine one, where we should not use the timeout for the streaming endpoints? So it could look just like this:
`private fun Channel.createStub() = CompactTxStreamerGrpcKt.CompactTxStreamerCoroutineStub(this, CallOptions.DEFAULT)`

Reply:
I hope I had an answer! 😂  I always found that the timeout api was not doing what I intended it to do. 

Just waving my hands to bring the attention to it 😅